### PR TITLE
Abstract Staging Textures - VideoCommon interface for texture readbacks/uploads

### PR DESCRIPTION
**Pull request by stenzek**
This PR creates an interface in VideoCommon for efficiently uploading/downloading to AbstractTextures.

Both uploads and downloads can be done asynchronously, with the interface managing fences and barriers. However, it is mainly beneficial for downloads. You can kick the copy off a frame or two earlier, and then read it back later with (hopefully) no idleing of the GPU. The first iteration of this change was included in the locking PR.

In very brief testing: roughly 10% faster for OpenGL frame dumping prior to hybrid XFB being merged, and a 50% improvement after hybrid XFB was merged (as the async frame dumping was dropped).

### CODE REVIEW

File: Source/Core/VideoBackends/D3D/DXTexture.cpp [hunk 1]
```diff
@@ -93,97 +96,36 @@ void DXTexture::Bind(unsigned int stage)
...
-  if (FAILED(hr))
-  {
-    WARN_LOG(VIDEO, "Failed to map texture dumping readback texture: %X", static_cast<u32>(hr));
-    return {};
-  }
-
-  return AbstractTexture::RawTextureInfo{reinterpret_cast<u8*>(map.pData), map.RowPitch,
-                                         m_config.width, m_config.height};
+  const DXTexture* srcentry = static_cast<const DXTexture*>(src);
+  _assert_(src_rect.GetWidth() == dst_rect.GetWidth() &&
```
Comment (iwubcode) [inline Source/Core/VideoBackends/D3D/DXTexture.cpp]: Why isn't this allowed?  Are you planning on having a separate function `CopyTexture` instead?

Comment (iwubcode) [inline Source/Core/VideoBackends/D3D/DXTexture.cpp]: If the source matches the destination, why not use `CopyResource` below??  I'd imagine it is faster.

Also, why wouldn't you want to allow copying a bit of a source into the same size or larger destination?

File: Source/Core/VideoBackends/D3D/DXTexture.cpp [hunk 2]
```diff
@@ -213,8 +155,140 @@ void DXTexture::CopyRectangleFromTexture(const AbstractTexture* source,
...
+  _assert_(src_rect.left >= 0 && static_cast<u32>(src_rect.right) <= src->GetConfig().width &&
+           src_rect.top >= 0 && static_cast<u32>(src_rect.bottom) <= src->GetConfig().height);
+  _assert_(dst_rect.left >= 0 && static_cast<u32>(dst_rect.right) <= m_config.width &&
+           dst_rect.top >= 0 && static_cast<u32>(dst_rect.bottom) <= m_config.height);
+
+  if (IsMapped())
+    DXStagingTexture::Unmap();
+
+  CD3D11_BOX src_box(src_rect.left, src_rect.top, 0, src_rect.right, src_rect.bottom, 1);
+  D3D::context->CopySubresourceRegion(
```
Comment (iwubcode) [inline Source/Core/VideoBackends/D3D/DXTexture.cpp]: Again, I believe you can use `CopyRegion` here

File: Source/Core/VideoBackends/OGL/OGLTexture.cpp [hunk 3]
```diff
@@ -315,4 +302,252 @@ void OGLTexture::SetStage()
...
+                 GetGLFormatForTextureFormat(m_config.format),
+                 GetGLTypeForTextureFormat(m_config.format), reinterpret_cast<void*>(dst_offset));
+
+    // Reset both read/draw framebuffers.
+    glBindFramebuffer(GL_FRAMEBUFFER, FramebufferManager::GetEFBFramebuffer());
+  }
+  else
+  {
+    // TODO: Investigate whether it's faster to use glReadPixels() with a framebuffer, since we're
+    // copying the whole texture, which may waste bandwidth. So we're trading CPU work in creating
```
Comment (iwubcode) [inline Source/Core/VideoBackends/OGL/OGLTexture.cpp]: Why not leverage `glGetTextureSubImage` ?

Comment (stenzek) [reply Source/Core/VideoBackends/OGL/OGLTexture.cpp]: Could use that, assuming GL4.5 is supported.

File: Source/Core/VideoBackends/D3D/DXTexture.cpp (hunk 1 above)
Comment (stenzek) [reply Source/Core/VideoBackends/D3D/DXTexture.cpp]: CopyResource will copy all array slices/mip levels, which may not be what we want. Especially if you want to copy level 1 of the source texture to level 0 of the destination texture. The dimensions have to match to do a blit/memcpy copy, if they don't match, that's why I have the Scale method.

File: Source/Core/VideoBackends/D3D/DXTexture.cpp (hunk 2 above)
Comment (stenzek) [reply Source/Core/VideoBackends/D3D/DXTexture.cpp]: Same deal as above. Staging textures at this point are designed to only be a single mip level.

**Comment by stenzek** [comment]
Thanks, I'll look into the D3D crash.

**Comment by stenzek** [comment]
D3D crashes should be fixed. Also moved the remaining statics to the Renderer class. We should really use smart pointers for these, but that's a larger change.

File: Source/Core/VideoBackends/D3D/DXTexture.cpp (hunk 1 above)
Comment (iwubcode) [reply Source/Core/VideoBackends/D3D/DXTexture.cpp]: Ah, totally missed the fact you differentiate between dst/src layer and level.  Thank you for the explanation!

**Comment by iwubcode** [comment]
@stenzek - can confirm that the D3D crash is fixed.  Thanks!  However, still seeing errors with dumping using `fifo_comparer` where no frames/avi seems to be outputted.  Seems like fifoci fairs a little better but there's quite a few failures:  https://fifoci.dolphin-emu.org/version/2d3f2a5b33108ad61ae19611be2902a4e50210e1/

One thing I noticed was the final frame isn't being outputted.  That is something I ran into too but can't recall how I fixed it.

**Comment by stenzek** [comment]
@iwubcode the last frame should be written now. Thanks for picking up on it.

**Comment by stenzek** [comment]
Hmm, looks like there might be an issue with the fences in Mesa, based on the fifoci results having partially rendered frames. I'll look into it.

**Comment by iwubcode** [comment]
Did a quick `fifo_comparer` test against a handful of logs.  Everything seems to be working!  Nice job.

**Comment by stenzek** [comment]
Mesa issues should be fixed now. Also fixes a bunch of regressions in fifoci from Hybrid XFB, e.g.

https://fifoci.dolphin-emu.org/compare/3356598-3355993/
https://fifoci.dolphin-emu.org/compare/3356582-3355942/
https://fifoci.dolphin-emu.org/compare/3356650-3355992/

**Comment by iwubcode** [comment]
@stenzek - nice!  Not seeing what you changed though to fix those regressions can you recall what you might have fixed?

Btw, code LGTM from me.  Excellent job!

**Comment by iwubcode** [comment]
@degasus - any concerns here or can this be merged?

**Pull request merged**